kong/plugins/grpc-gateway/deco.lua

```diff
@@ -51,165 +67,207 @@ local valid_method = {
-  local match_groups = {}
-  local match_group_idx = 1
-  local path_regex, _, err = re_gsub("^" .. path .. "$", options_path_regex, function(m)
+--- Parse path in http.options
+-- @param request path
+-- @return Regular expression to extract variables from request path
+-- @return Variables found in request path
+-- @return Error string
+local function parse_options_path( path )
+  local path_vars = {}
+
+  local path_regex, _, err = re_gsub( "^" .. path .. "$", options_path_regex, function(m)
     local var = m[1]
     local paths = m[2]
     -- store lookup table to matched groups to variable name
-    match_groups[match_group_idx] = var
-    match_group_idx = match_group_idx + 1
-    if not paths or paths == "*" then
+    table.insert( path_vars, var )
+
+    if ( not paths ) or ( paths == "*" ) then
       return "([^/]+)"
     else
       return ("(%s)"):format(
-        paths:gsub("%*%*", ".+"):gsub("%*", "[^/]+")
+        paths:gsub( "%*%*", ".+" ):gsub( "%*", "[^/]+" )
       )
     end
   end, "jo")
+
   if err then
     return nil, nil, err
   end
 
-  return path_regex, match_groups
+  return path_regex, path_vars
 end
 
+--- Get information about call
+-- @param cfg Call configuration
+-- @return information about call
+local function get_call_info( cfg )
+  local filename = cfg.proto
 
--- parse, compile and load .proto file
--- returns a table mapping valid request URLs to input/output types
-local _proto_info = {}
-local function get_proto_info(fname)
-  local info = _proto_info[fname]
-  if info then
-    return info
+  if get_call_info_cache[ filename ] then
+    return get_call_info_cache[ filename ]
   end
 
-  info = {}
+  local info = {
+    endpoints = {},
+    json_names = {}
+  }
 
-  local grpc_tools_instance = grpc_tools.new()
-  grpc_tools_instance:each_method(fname, function(parsed, srvc, mthd)
-    local options_bindings =  {
-      safe_access(mthd, "options", "options", "google.api.http"),
-      safe_access(mthd, "options", "options", "google.api.http", "additional_bindings")
+  local load_bindings_info = function( file, service, method )
+    local bindings =  {
+      safe_access( method, "options", "options", "google.api.http"),
+      safe_access( method, "options", "options", "google.api.http", "additional_bindings")
     }
-    for _, options in ipairs(options_bindings) do
-      for http_method, http_path in pairs(options) do
+
+    for _, binding in ipairs( bindings ) do
+      for http_method, http_path in pairs( binding ) do
+
         http_method = http_method:lower()
-        if valid_method[http_method] then
-          local preg, grp, err = parse_options_path(http_path)
+
+        if is_valid_method[ http_method ] then
+          local regex, varnames, err = parse_options_path( http_path )
+
           if err then
-            ngx.log(ngx.ERR, "error ", err, "parsing options path ", http_path)
+            ngx.log( ngx.ERR, "error: ", err, " parsing options path: ", http_path )
```

Comment:
Usually, we don't add space after and before parentheses. I'm not commenting on all places.

```suggestion
            ngx.log(ngx.ERR, "error: ", err, " parsing options path: ", http_path)
```

Reply:
So for the brackets.